File: src/main/java/com/irisine/jiris/compiler/expression/IrisMemberExpression.java
package com.irisine.jiris.compiler.expression;

import com.irisine.jiris.compiler.IrisCompiler;
import com.irisine.jiris.compiler.IrisGenerateHelper;
import com.irisine.jiris.compiler.assistpart.IrisIdentifier;
import com.irisine.jiris.compiler.statement.IrisStatement;
import net.bytebuddy.dynamic.DynamicType;
import net.bytebuddy.jar.asm.MethodVisitor;
import org.irislang.jiris.compiler.IrisNativeJavaClass;

/**
 * Created by Huisama on 2017/5/2 0002.
 */
public class IrisMemberExpression extends IrisExpression {

    private IrisExpression m_caller = null;
    private IrisIdentifier m_property = null;

    public IrisMemberExpression(IrisExpression caller, IrisIdentifier property) {
        m_caller = caller;
        m_property = property;
    }

    @Override
    public boolean Generate(IrisCompiler currentCompiler, DynamicType.Builder<IrisNativeJavaClass> currentBuilder, MethodVisitor visitor) {
        IrisGenerateHelper.SetLineNumber(visitor, currentCompiler, getLineNumber());

        String property = m_property.getIdentifier();
        StringBuilder builder = new StringBuilder("__get_").append(property);

        if(!m_caller.Generate(currentCompiler, currentBuilder, visitor)) {
            return false;
        }

        IrisGenerateHelper.CallMethod(visitor, currentCompiler, builder.toString(), 0, false);

        return true;
    }

    @Override
    public LeftValueResult LeftValue(IrisCompiler currentCompiler, DynamicType.Builder<IrisNativeJavaClass> currentBuilder, MethodVisitor visitor) {
        LeftValueResult result = new LeftValueResult();
        String property = m_property.getIdentifier();
        StringBuilder builder = new StringBuilder("__set_").append(property);

        if(!m_caller.Generate(currentCompiler, currentBuilder, visitor)){
            result.setResult(false);
            return result;
        }

        result.setIdentifier(builder.toString());
        result.setResult(true);
        result.setType(LeftValueType.MemberVariable);
        return result;
    }
}

File: src/main/java/com/irisine/jiris/compiler/expression/IrisMemberExpression.kt
package com.irisine.jiris.compiler.expression

import com.irisine.jiris.compiler.IrisCompiler
import com.irisine.jiris.compiler.IrisGenerateHelper
import com.irisine.jiris.compiler.assistpart.IrisIdentifier
import net.bytebuddy.dynamic.DynamicType
import net.bytebuddy.jar.asm.MethodVisitor
import org.irislang.jiris.IrisNativeJavaClass

/**
 * Created by Huisama on 2017/5/2 0002.
 */
class IrisMemberExpression(val caller: IrisExpression, val property: IrisIdentifier) : IrisExpression() {

    override fun Generate(currentCompiler: IrisCompiler, currentBuilder: DynamicType.Builder<IrisNativeJavaClass>, visitor: MethodVisitor): Boolean {
        IrisGenerateHelper.SetLineNumber(visitor, currentCompiler, lineNumber)

        val property = property.identifier
//        val builder = StringBuilder("__get_").append(property)

        val funcName = "__get_${property}"

        if (!caller.Generate(currentCompiler, currentBuilder, visitor)) {
            return false
        }

        IrisGenerateHelper.CallMethod(visitor, currentCompiler, funcName, 0, false)

        return true
    }

    override fun LeftValue(currentCompiler: IrisCompiler, currentBuilder: DynamicType.Builder<IrisNativeJavaClass>, visitor: MethodVisitor): IrisExpression.LeftValueResult {
        val result = IrisExpression.LeftValueResult()
        val property = property.identifier
//        val builder = StringBuilder("__set_").append(property)

        val funcName = "__set_${property}"

        if (!caller.Generate(currentCompiler, currentBuilder, visitor)) {
            result.result = false
            return result
        }

        result.identifier = funcName
        result.result = true
        result.type = IrisExpression.LeftValueType.MemberVariable
        return result
    }
}
